src/plugins/intel_gpu/src/plugin/infer_request.cpp

```diff
@@ -620,21 +619,58 @@ Blob::Ptr InferRequest::create_shared_device_blob(const InferenceEngine::TensorD
 
 void InferRequest::copy_output_data(cldnn::memory::ptr src, Blob::Ptr dst) {
     OV_ITT_SCOPED_TASK(itt::domains::intel_gpu_plugin, "InferRequest::copy_output_data");
+    auto is_convert_needed = [](const Precision& prc) {
+        const std::vector<Precision> convert_needed = { Precision::I16, Precision::U16, Precision::FP64,
+                                                        Precision::U32, Precision::U64 };
+        const std::vector<Precision> convert_not_needed = { Precision::FP32, Precision::FP16, Precision::I64, Precision::I32,
+                                                            Precision::I8, Precision::U8, Precision::BOOL };
+
+        if (std::find(convert_needed.begin(), convert_needed.end(), prc) != convert_needed.end())
+            return true;
+        else if (std::find(convert_not_needed.begin(), convert_not_needed.end(), prc) != convert_not_needed.end())
+            return false;
+        else
+            OPENVINO_ASSERT(false, "[GPU] Plugin does not support output ", prc, " precision");
+    };
+
+    const auto convert_needed = is_convert_needed(dst->getTensorDesc().getPrecision());
+    const auto src_layout = src->get_layout();
     auto& stream = m_graph->GetNetwork()->get_stream();
-    switch (dst->getTensorDesc().getPrecision()) {
-    case Precision::FP64: copyResultToOutputBlob<float, double>(src, dst, stream);    break;
-    case Precision::FP32: copyResultToOutputBlob<float, float>(src, dst, stream);    break;
-    case Precision::FP16: copyResultToOutputBlob<uint16_t, uint16_t>(src, dst, stream); break;
-    case Precision::I64:  copyResultToOutputBlob<int64_t, int64_t>(src, dst, stream);  break;
-    case Precision::I32:  copyResultToOutputBlob<int32_t, int32_t>(src, dst, stream);  break;
-    case Precision::I16:  copyResultToOutputBlob<float, int16_t>(src, dst, stream);  break;
-    case Precision::I8:   copyResultToOutputBlob<int8_t, int8_t>(src, dst, stream);  break;
-    case Precision::U16:  copyResultToOutputBlob<float, uint16_t>(src, dst, stream);  break;
-    case Precision::U32:  copyResultToOutputBlob<int32_t, uint32_t>(src, dst, stream);  break;
-    case Precision::U64:  copyResultToOutputBlob<int32_t, uint64_t>(src, dst, stream);  break;
-    case Precision::U8:   copyResultToOutputBlob<uint8_t, uint8_t>(src, dst, stream);  break;
-    case Precision::BOOL: copyResultToOutputBlob<int8_t, int8_t>(src, dst, stream);  break;
-    default: IE_THROW(NotImplemented) << "The plugin does not support output " << dst->getTensorDesc().getPrecision() << " precision";
+
+    if (convert_needed || src_layout.data_padding) {
+        static thread_local Blob::Ptr intermediate_blob = nullptr;
```

Comment:
nit: maybe it's better to make it field of InferRequest class? Also, please add a comment on why it's needed